schema/deploy/database_functions/validate_energy_products.sql

```diff
@@ -0,0 +1,56 @@
+-- Deploy ggircs-portal:database_functions/validate_energy_products to pg
+-- requires: views/ciip_production
+
+begin;
+
+create or replace function ggircs_portal_private.validate_energy_products(
+    product_array ggircs_portal.ciip_production[]
+  )
+    returns void
+    as
+    $function$
+    declare
+      product record;
+      product_data ggircs_portal.product;
+      product_ids int[];
+      i int;
+    begin
+      i=1;
+      foreach product in array product_array
+        loop
+          product_ids[i] = product.product_id;
+          i = i + 1;
+        end loop;
+
+      foreach product in array product_array
+        loop
+
+          select * into product_data from ggircs_portal.product
+          where id = product.product_id;
+
+          case
+            when (product_data.add_purchased_electricity_emissions and (select not 3 = any (product_ids))) then
+              raise exception 'Reported product requires purchased_electricity to be reported, but purchased_electricity is missing';
+            when (product_data.subtract_exported_electricity_emissions and (select not 1 = any (product_ids))) then
+              raise exception 'Reported product requires exported_electricity to be reported, but exported_electricity is missing';
+            when (product_data.add_purchased_heat_emissions and (select not 4 = any (product_ids))) then
+              raise exception 'Reported product requires purchased_heat to be reported, but purchased_heat is missing';
+            when (product_data.subtract_exported_heat_emissions and (select not 2 = any (product_ids))) then
+              raise exception 'Reported product requires exported_heat to be reported, but exported_heat is missing';
+            when (product_data.subtract_generated_electricity_emissions and (select not 5 = any (product_ids))) then
+              raise exception 'Reported product requires generated_electricity to be reported, but generated_electricity is missing';
+            when (product_data.subtract_generated_heat_emissions and (select not 6 = any (product_ids))) then
+              raise exception 'Reported product requires generated_heat to be reported, but generated_heat is missing';
+            when (product_data.add_emissions_from_eios and (select not 7 = any (product_ids))) then
+              raise exception 'Reported product requires emissions_from_eios to be reported, but emissions_from_eios is missing';
```

Comment:
The `application_revision_ciip_incentive_test.sql` is truncating and setting these ids directly. How do we guarantee we have the same ids in dev, test, and prod?

Reply:
Yes, these exceptions are tightly coupled to the ids. I'd love to hear any suggestions on not doing this, but I can't think of a way that isn't tightly coupled to either the name of the energy product, or the id. 

What we've done with these energy products is they are created from schema/data/prod/energy_products.sql, so they are always created first, and always created the same in dev test and prod on deploy. I've also added a comment in that file (in my subsequent PR) that mentions that if these change at all there are cascading effects to this sql file as well as the front-end validation. These products are protected by a trigger that does not allow them to be editable from the UI or on the database level. The trigger would have to be disabled to make any changes.

Reply:
could we add a required_for column somewhere? Is the requirement relation a one-to-many or a many-to-many?

Reply:
Since you've got a mitigation in place, this might be trackable as a tech debt card, enabling merge now and reassessment later?

Reply:
The energy_products are required for the incentive calculation to work, any non-energy product can require none, some or all of these energy_products to be reported along with it (as defined by the admin), so I suppose many to many? In which case a required_for column wouldn't be a good option, a mapping table could be a better solution. Since the mitigation in place at the moment seems pretty solid I'd rather get this merged & create a tech debt card to investigate rather than muck with mapping tables right now